src/rust/src/buf.rs

```diff
@@ -46,47 +40,130 @@ fn _extract_buffer_length<'p>(
-        let (bufobj, ptrval) = _extract_buffer_length(pyobj, false)?;
-        let len = bufobj.len()?;
-        let buf = if len == 0 {
-            &[]
-        } else {
-            // SAFETY: _extract_buffer_length ensures that we have a valid ptr
-            // and length (and we ensure we meet slice's requirements for
-            // 0-length slices above), we're keeping pyobj alive which ensures
-            // the buffer is valid. But! There is no actually guarantee
-            // against concurrent mutation. See
-            // https://alexgaynor.net/2022/oct/23/buffers-on-the-edge/
-            // for details. This is the same as our cffi status quo ante, so
-            // we're doing an unsound thing and living with it.
-            unsafe { slice::from_raw_parts(ptrval as *const u8, len) }
-        };
+#[cfg(not(Py_3_11))]
+mod ffi_impl {
+    use super::_extract_buffer_length;
+    use pyo3::types::PyAnyMethods;
+    use std::slice;
 
-        Ok(CffiBuf {
-            pyobj: pyobj.clone(),
-            _bufobj: bufobj,
-            buf,
-        })
+    pub(crate) struct CffiBuf<'p> {
+        pyobj: pyo3::Bound<'p, pyo3::PyAny>,
+        _bufobj: pyo3::Bound<'p, pyo3::PyAny>,
+        buf: &'p [u8],
+    }
+
+    impl<'a> CffiBuf<'a> {
+        pub(crate) fn from_bytes(py: pyo3::Python<'a>, buf: &'a [u8]) -> Self {
+            CffiBuf {
+                pyobj: py.None().into_bound(py),
+                _bufobj: py.None().into_bound(py),
+                buf,
+            }
+        }
+
+        pub(crate) fn as_bytes(&self) -> &[u8] {
+            self.buf
+        }
+
+        pub(crate) fn into_pyobj(self) -> pyo3::Bound<'a, pyo3::PyAny> {
+            self.pyobj
+        }
+    }
+
+    impl<'a> pyo3::conversion::FromPyObject<'a> for CffiBuf<'a> {
+        fn extract_bound(pyobj: &pyo3::Bound<'a, pyo3::PyAny>) -> pyo3::PyResult<Self> {
+            let (bufobj, ptrval) = _extract_buffer_length(pyobj, false)?;
+            let len = bufobj.len()?;
+            let buf = if len == 0 {
+                &[]
+            } else {
+                // SAFETY: _extract_buffer_length ensures that we have a valid ptr
+                // and length (and we ensure we meet slice's requirements for
+                // 0-length slices above), we're keeping pyobj alive which ensures
+                // the buffer is valid. But! There is no actually guarantee
+                // against concurrent mutation. See
+                // https://alexgaynor.net/2022/oct/23/buffers-on-the-edge/
+                // for details. This is the same as our cffi status quo ante, so
+                // we're doing an unsound thing and living with it.
+                unsafe { slice::from_raw_parts(ptrval as *const u8, len) }
+            };
+
+            Ok(CffiBuf {
+                pyobj: pyobj.clone(),
+                _bufobj: bufobj,
+                buf,
+            })
+        }
     }
 }
```

Comment:
`CffiMutBuf` was not touched as `get` is implemented as `ffi::PyObject_GetBuffer(obj.as_ptr(), buf.as_mut_ptr(), ffi::PyBUF_FULL_RO)` so it would happily accept `PyBytes` objects and write into them.  I wasn't sure what kind of guard was needed there and if something like `obj.is_instance_of::<pyo3::types::PyBytes>` would have been enough so I did not implement it but likely could be done in a future PR.

Reply:
I believe all that's required is checking the `readonly` property on the `Buffer` object and raising if its true.

Reply:
I had tried that first but `PyBytes` is not `readonly`

Reply:
wat... how can that possibly be right?

Reply:
I can't reproduce this:

```
fn main() -> anyhow::Result<()> {
    pyo3::prepare_freethreaded_python();

    pyo3::Python::with_gil(|py| -> pyo3::PyResult<()> {
        let b = pyo3::types::PyBytes::new(py, b"hello world");
        let buf = pyo3::buffer::PyBuffer::<u8>::get(&b)?;
        dbg!(buf.readonly());

        let b = pyo3::types::PyByteArray::new(py, b"hello world");
        let buf = pyo3::buffer::PyBuffer::<u8>::get(&b)?;
        dbg!(buf.readonly());

        Ok(())
    })?;

    Ok(())
}
```

Reply:
Err, forgot to include the result:

```
[src/main.rs:7:9] buf.readonly() = true
[src/main.rs:11:9] buf.readonly() = false
```

Reply:
This test will fail: https://github.com/pyca/cryptography/blob/7a49948ebf22bc8050cd34bc6cce7154634fadc8/tests/hazmat/primitives/test_ciphers.py#L231. It looks like the `readonly` flag isn’t automatically derived from Python-level mutability, objects like bytes are immutable to Python, but the buffer interface doesn’t appear to reflect that.

Reply:
Let me try it again. Hopefully I got something wrong the first time

Reply:
You're right, I must've messed something up on my first try. I just tested it again and it works exactly as you described.

Pushed in 6bf9f981c885171d5d29d2cd86bb8a1780ac0a20. I'll follow up with some cleanup to make it more DRY.